### PR TITLE
COMPAT: Consider Python 2.x tarfiles file-like

### LINKED ISSUE #16530: pandas read_csv no longer supports file-like objects from tarfile (pandas 0.20.1)
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import tarfile

tar = tarfile.open(name="xxx.tar.bz2", mode='r')
myfile = tar.extractfile('yyy.csv') # file-like object with a read() method
data = pd.read_csv(myfile, sep=r'\s+')
```
This code generates this error:
```
  File "/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.py", line 655, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.py", line 392, in _read
    filepath_or_buffer, encoding, compression)
  File "/usr/local/lib/python2.7/dist-packages/pandas/io/common.py", line 210, in get_filepath_or_buffer
    raise ValueError(msg.format(_type=type(filepath_or_buffer)))
ValueError: Invalid file path or buffer object type: <class 'tarfile.ExFileObject'>
```

#### Problem description

This code works with pandas 0.19.2 but fails with 0.20.1.

According to pandas doc for read_csv:

> filepath_or_buffer : str, pathlib.Path, py._path.local.LocalPath or any object with a read() method (such as a file handle or StringIO)

I guess the new validations are too restrictive ?

### PR DESCRIPTION
**Pull request by gfyoung**
`Tarfile.ExFileObject` has no `next` method in Python 2.x, making it an invalid file-like object in `read_csv`. However, they can be read in just fine, meaning our check is too strict for file-like. This commit relaxes
the check to just look for the `__iter__` attribute.

Closes #16530.

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -801,6 +803,23 @@ def _get_options_with_defaults(self, engine):
 
         return options
 
+    def _check_file_or_buffer(self, f, engine):
+        # see gh-16530
+        if is_file_like(f):
+            next_attr = "__next__" if PY3 else "next"
+
+            if engine != "c" and not hasattr(f, next_attr):
```
Comment (jreback) [inline pandas/io/parsers.py]: can you add a comment of why you are doing this

File: setup.py [hunk 2]
```diff
@@ -702,6 +702,7 @@ def pxd(name):
                                         'parser/data/*.gz',
                                         'parser/data/*.bz2',
                                         'parser/data/*.txt',
+                                        'parser/data/*.tar',
```
Comment (jreback) [inline setup.py]: can you add with ``.tar.gz`` as well?

**Review by jreback — APPROVED** [review]
lgtm. minor comments.

File: setup.py (hunk 2 above)
Comment (gfyoung) [inline setup.py]: Done.

File: pandas/io/parsers.py (hunk 1 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -801,6 +803,27 @@ def _get_options_with_defaults(self, engine):
...
+        # see gh-16530
+        if is_file_like(f):
+            next_attr = "__next__" if PY3 else "next"
+
+            # The C engine doesn't need the file-like to have
+            # the "__next__" attribute. However, the Python
+            # explicitly calls "next(...)" when iterating through
+            # such an object, meaning it needs to have that attribute.
+            if engine != "c" and not hasattr(f, next_attr):
+                msg = ("The 'python' engine can not iterate "
```
Comment (TomAugspurger) [inline pandas/io/parsers.py]: this block seems untested. Are you able to add some, or is it a hassle to hit it?

File: setup.py (hunk 2 above)
Comment (jreback) [inline setup.py]: I meant test with ``.tar.gz`` as well.

**Review by jreback — APPROVED** [review]
lgtm. couple of comments.

File: pandas/io/parsers.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Was meaning to do that eventually.  Done.

File: setup.py (hunk 2 above)
Comment (gfyoung) [inline setup.py]: Ah, I see.  Done.

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by gfyoung** [comment]
@jreback : Green all around and ready to merge.

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -801,6 +803,28 @@ def _get_options_with_defaults(self, engine):
...
+            # for Python 3.x)
+            if engine != "c" and not hasattr(f, next_attr):
+                msg = ("The 'python' engine cannot iterate "
+                       "through this file buffer.")
+                if self._engine_specified:
+                    raise ValueError(msg)
+                else:
+                    engine = "c"
+                    msg += " Falling back to the 'c' engine."
+                    warnings.warn(msg, ParserWarning, stacklevel=5)
```
Comment (bashtage) [inline pandas/io/parsers.py]: Appears this isn't being hit by test coverage.  Should the warning be checked?  Can this actually be hit?

Comment (gfyoung) [inline pandas/io/parsers.py]: You're right.  This actually can't be hit.  Removing.

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -801,6 +803,24 @@ def _get_options_with_defaults(self, engine):
 
         return options
 
+    @staticmethod
```
Comment (jreback) [inline pandas/io/parsers.py]: why is this a staticmethod?

Comment (gfyoung) [inline pandas/io/parsers.py]: Because `self.` is not used.  I pass in the `engine` and `f` as parameters.

Comment (jreback) [inline pandas/io/parsers.py]: right but no other methods are like this. you can make it a free function if you want, but prob just make it a method.

Comment (gfyoung) [inline pandas/io/parsers.py]: Alright.  Done.

**Comment by jreback** [comment]
thanks!

**Pull request merged**